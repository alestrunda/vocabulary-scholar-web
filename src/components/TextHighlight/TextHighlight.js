/*global window*/

import React from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { splitString } from '../../misc';

const TextHightlight = props => {
  const handleClick = () => {
    //check if user selected any part of the string passed as children
    if (window.getSelection().rangeCount) {
      //get indexes of the selected part
      const range = window.getSelection().getRangeAt(0);
      props.onHighlight(range.startOffset, range.endOffset);
    }
  };

  const { start, middle, end } = splitString(
    props.children,
    props.start,
    props.end
  );

  return (
    <div onClick={handleClick} className={'highlighting-text'}>
      <div className={classNames('highlighting-text__text', props.className)}>
        {props.children}
      </div>
      {middle && (
        <div className="highlighting-text__over">
          <span data-testid="part-1">{start}</span>
          <span data-testid="part-2" className={props.classNameHighlighted}>
            {middle}
          </span>
          <span data-testid="part-3">{end}</span>
        </div>
      )}
    </div>
  );
};

TextHightlight.propTypes = {
  className: PropTypes.string,
  classNameHighlighted: PropTypes.string,
  children: PropTypes.string,
  onHighlight: PropTypes.func.isRequired,
  start: PropTypes.number,
  end: PropTypes.number,
};

TextHightlight.defaultProps = {
  classNameHighlighted: 'text-highlight',
  children: '',
};

export default TextHightlight;
