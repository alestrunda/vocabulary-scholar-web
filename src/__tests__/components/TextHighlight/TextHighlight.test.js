import React from 'react';
import { shallow } from 'enzyme';

import TextHighlight from '../../../components/TextHighlight';

it('renders correctly', () => {
  const onHighlight = jest.fn();
  const root = shallow(
    <TextHighlight onHighlight={onHighlight}>Lorem Ipsum</TextHighlight>
  );
  expect(root).toMatchSnapshot();
});

it('renders correctly highlighted part', () => {
  const onHighlight = jest.fn();
  const root = shallow(
    <TextHighlight start={0} end={5} onHighlight={onHighlight}>
      Lorem Ipsum
    </TextHighlight>
  );
  expect(root).toMatchSnapshot();
});
