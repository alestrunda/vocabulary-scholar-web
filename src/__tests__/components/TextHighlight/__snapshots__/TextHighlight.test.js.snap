// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextHightlight
    classNameHighlighted="text-highlight"
    onHighlight={[MockFunction]}
  >
    Lorem Ipsum
  </TextHightlight>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="highlighting-text__text"
        >
          Lorem Ipsum
        </div>,
        undefined,
      ],
      "className": "highlighting-text",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Lorem Ipsum",
          "className": "highlighting-text__text",
        },
        "ref": null,
        "rendered": "Lorem Ipsum",
        "type": "div",
      },
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="highlighting-text__text"
          >
            Lorem Ipsum
          </div>,
          undefined,
        ],
        "className": "highlighting-text",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Lorem Ipsum",
            "className": "highlighting-text__text",
          },
          "ref": null,
          "rendered": "Lorem Ipsum",
          "type": "div",
        },
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;

exports[`renders correctly highlighted part 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <TextHightlight
    classNameHighlighted="text-highlight"
    end={5}
    onHighlight={[MockFunction]}
    start={0}
  >
    Lorem Ipsum
  </TextHightlight>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="highlighting-text__text"
        >
          Lorem Ipsum
        </div>,
        <div
          className="highlighting-text__over"
        >
          <span
            data-testid="part-1"
          >
            
          </span>
          <span
            className="text-highlight"
            data-testid="part-2"
          >
            Lorem
          </span>
          <span
            data-testid="part-3"
          >
             Ipsum
          </span>
        </div>,
      ],
      "className": "highlighting-text",
      "onClick": [Function],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "Lorem Ipsum",
          "className": "highlighting-text__text",
        },
        "ref": null,
        "rendered": "Lorem Ipsum",
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <span
              data-testid="part-1"
            >
              
            </span>,
            <span
              className="text-highlight"
              data-testid="part-2"
            >
              Lorem
            </span>,
            <span
              data-testid="part-3"
            >
               Ipsum
            </span>,
          ],
          "className": "highlighting-text__over",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "",
              "data-testid": "part-1",
            },
            "ref": null,
            "rendered": "",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Lorem",
              "className": "text-highlight",
              "data-testid": "part-2",
            },
            "ref": null,
            "rendered": "Lorem",
            "type": "span",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": " Ipsum",
              "data-testid": "part-3",
            },
            "ref": null,
            "rendered": " Ipsum",
            "type": "span",
          },
        ],
        "type": "div",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="highlighting-text__text"
          >
            Lorem Ipsum
          </div>,
          <div
            className="highlighting-text__over"
          >
            <span
              data-testid="part-1"
            >
              
            </span>
            <span
              className="text-highlight"
              data-testid="part-2"
            >
              Lorem
            </span>
            <span
              data-testid="part-3"
            >
               Ipsum
            </span>
          </div>,
        ],
        "className": "highlighting-text",
        "onClick": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Lorem Ipsum",
            "className": "highlighting-text__text",
          },
          "ref": null,
          "rendered": "Lorem Ipsum",
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <span
                data-testid="part-1"
              >
                
              </span>,
              <span
                className="text-highlight"
                data-testid="part-2"
              >
                Lorem
              </span>,
              <span
                data-testid="part-3"
              >
                 Ipsum
              </span>,
            ],
            "className": "highlighting-text__over",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "",
                "data-testid": "part-1",
              },
              "ref": null,
              "rendered": "",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Lorem",
                "className": "text-highlight",
                "data-testid": "part-2",
              },
              "ref": null,
              "rendered": "Lorem",
              "type": "span",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": " Ipsum",
                "data-testid": "part-3",
              },
              "ref": null,
              "rendered": " Ipsum",
              "type": "span",
            },
          ],
          "type": "div",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromError": true,
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    Symbol(enzyme.__providerValues__): undefined,
  },
  Symbol(enzyme.__providerValues__): Map {},
}
`;
